File: backend/app/infrastructure/models/finance.py
from beanie import Document
from pydantic import Field
from typing import Optional, List
from datetime import datetime
from enum import Enum

class PaymentMethodEnum(str, Enum):
    MOBILE_MONEY = "mobile_money"
    BANK_TRANSFER = "bank_transfer"
    CARD = "card"
    CASH = "cash"

class PaymentStatusEnum(str, Enum):
    PENDING = "pending"
    SUCCESS = "success"
    FAILED = "failed"
    CANCELLED = "cancelled"

class FeeStructure(Document):
    tenant_id: str
    programme_id: Optional[str] = None
    level: Optional[str] = None
    academic_year: str

    fees: dict = {}

    created_at: datetime = Field(default_factory=datetime.utcnow)

    class Settings:
        name = "fee_structures"

class Payment(Document):
    tenant_id: str
    student_id: Optional[str] = None
    applicant_id: Optional[str] = None

    amount: float
    fee_type: str
    academic_year: Optional[str] = None

    payment_method: PaymentMethodEnum
    payment_reference: str
    status: PaymentStatusEnum = PaymentStatusEnum.PENDING

    paystack_reference: Optional[str] = None

    payment_date: Optional[datetime] = None
    receipt_number: Optional[str] = None
    receipt_pdf_url: Optional[str] = None

    created_at: datetime = Field(default_factory=datetime.utcnow)

    class Settings:
        name = "payments"
        indexes = [
            [("tenant_id", 1), ("student_id", 1)],
            [("status", 1)],
            [("payment_date", 1)],
        ]

class Scholarship(Document):
    tenant_id: str
    student_id: str

    name: str
    scholarship_type: str
    amount: float
    percentage: Optional[float] = None

    approved_by: str
    approved_date: datetime

    start_date: datetime
    end_date: Optional[datetime] = None
    is_active: bool = True

    class Settings:
        name = "scholarships"

File: frontend/src/types/finance.ts
export interface FeeStructureCreateRequest {
  programme_id?: string
  level?: string
  academic_year: string
  fees: Record<string, number>
}

export interface FeeStructureResponse {
  id: string
  programme_id?: string
  level?: string
  academic_year: string
  fees: Record<string, number>
}

export interface ScholarshipRequest {
  student_id: string
  name: string
  scholarship_type: string
  amount: number
  percentage?: number
  start_date: string
  end_date?: string
}

export interface ScholarshipResponse {
  id: string
  student_id: string
  name: string
  scholarship_type: string
  amount: number
  percentage?: number
  start_date: string
  end_date?: string | null
  is_active: boolean
}

export interface PaymentListItem {
  id: string
  tenant_id: string
  student_id?: string | null
  applicant_id?: string | null
  amount: number
  fee_type: string
  academic_year?: string | null
  payment_method: string
  payment_reference: string
  status: string
  paystack_reference?: string | null
  payment_date?: string | null
  receipt_number?: string | null
  created_at: string
}

export interface BalanceResponse {
  total_due: number
  total_paid: number
  total_scholarships: number
  balance: number
}

export interface ClearanceResponse {
  student_id: string
  tenant_id: string
  is_cleared: boolean
  message: string
  balance: number
  pending_payments: number
}
